Add ROC metric

joanne-tsai opened the pull request:
Before there was no ROC metric. Having an ROC metric helps with visualizing the performance of the algorithm. This adds the ROC metric as well as some edge test cases to ensure that the ROC curve is calculated correctly.

joanne-tsai committed:
Add ROC metric

The ROC metric helps visualize the threshold that performs the best.
The macro average of the ROC curve is not calculated because it wouldn't be
helpful to have.

joanne-tsai committed:
Add ROC test cases

There are 4 test cases used: 1 taken from the sklearn website, a perfect
predictor, a 50/50 predictor, and a wrong predictor. For the sake of
simplicity, the test cases I wrote are only edge cases.

brianhelba commented on isic_challenge_scoring/metrics.py:
```diff
@@ -174,3 +174,9 @@ def auc_above_sensitivity(
+
+def roc(truth_probabilities: pd.Series, prediction_probabilities: pd.Series):
+    fpr, tpr, thresholds = sklearn.metrics.roc_curve(truth_probabilities, prediction_probabilities)
+    roc = list(map(lambda x, y, t: {'x': x, 'y': y, 'thresh': t}, fpr, tpr, thresholds))
```
Could we use keys of `tpr`, `fpr`, and `threshold` for this? X and Y are properties of the ROC rendering, rather than measurements of the algorithm.

brianhelba commented on isic_challenge_scoring/task3.py:
```diff
@@ -56,6 +56,7 @@ def compute_metrics(truth_file_stream, prediction_file_stream) -> Dict[str, Dict
             'ap': metrics.average_precision(
                 truth_category_probabilities, prediction_category_probabilities
             ),
+            # 'roc': metrics.roc(truth_category_probabilities, prediction_category_probabilities),
```
Why isn't this being used? Why do we need to set it in another loop below?

joanne-tsai replied:
If I put it in that loop, the macro_average tries to calculate the average of the roc curve and throws an error. I couldn't figure out how to make the macro_average skip roc so I thought to just add another loop at the end to add the roc metric in.

joanne-tsai committed:
Replace roc metric's keys with 'fpr', 'tpr', 'threshold'

x and y are properties of graphs whereas fpr and tpr are measurements on the
curve that is returned by sklearn.
'threshold' is a more proper name than 'thresh'

joanne-tsai committed:
Change the way roc is added into dict of metrics

Before roc was added in a separate loop to avoid conflicts with the
macro_average. An if statement was added to the macro_average to allow for roc
to be added in the same loop as the other metrics while not having any conflicts

joanne-tsai committed:
Fix linting errors

brianhelba reviewed (approved):
Looks great. I particularly like the tests.

For large inputs, this might be outputting too many ROC points, but we can address that if necessary in the future.
